cbmc/proofs/proof_guide.md

```diff
@@ -0,0 +1,399 @@
+
+This also explains why we prefer to use _signed_ integers for loop counters and quantified variables - to allow "0 - 1" to evaluate to -1.  If you use an unsigned type, then "0 - 1" evaluated to something like UINT_MAX, and all hell breaks loose.
+
+## Invariant && loop-exit ==> post-condition
+
+Another important sanity check. If there are no statements following the loop body, then the loop invariant AND the loop exit condition should imply the post-condition. For the example above, we need to prove:
+
+```
+// Loop invariant
+(i >= 0 &&
+ i <= len &&
+ (FORALL { int j; (0 <= j && j <= i - 1) ==> dst[j] == 0 } )
+&&
+// Loop exit condition must be TRUE, so
+i == len)
+
+  ==>
+
+// Post-condition
+FORALL { int k; (0 <= k && k < len) ==> dst[k] == 0 }
+```
+
+The loop exit condition means that we can just replace `i` with `len` in the hypotheses, to get:
+
+```
+len >= 0 &&
+len <= len &&
+(FORALL { int j; (0 <= j && j <= len - 1) ==> dst[j] == 0 } )
+  ==>
+FORALL { int k; (0 <= k && k < len) ==> dst[k] == 0 }
+```
+
+`j <= len - 1` can be simplified to `j < len` so that simplifies to:
+
+```
+FORALL { int j; (0 <= j && j < len) ==> dst[j] == 0 }
+  ==>
+FORALL { int k; (0 <= k && k < len) ==> dst[k] == 0 }
+```
+
+which is True.
+
+## Recipe to prove a new function
+
+Let's say we want to develop a proof of a function. Here are the basic steps.
+1. Populate a proof directory
+2. Update Makefile
+3. Update harness function
+4. Supply top-level contracts for the function
+5. Supply loop-invariants (if requied) and other interior contracts
+6. Prove it!
+
+These steps are expanded on in the following sub-sections
+
+### Populate a proof directory
+
+For MLKEM-C-AArch64, proof directories lie below `cbmc/proofs`
+
+Create a new sub-directory in there, where the name of the directory is the name of the function _without_ the `$(KYBER_NAMESPACE)` prefix.
+
+That directory needs to contain 3 files.
+
+* cbmc-proof.txt
+* Makefile
+* XXX_harness.c
+
+where "XXX" is the name of the function being proved - same as the directory name.
+
+I suggest that you copy these files from an existing proof directory and modify the latter two. The `cbmc-proof.txt` file is just a marker that says "this directory contains a CBMC proof" to the tools, so no modification is required.
+
+### Update Makefile
+
+The `Makefile` sets options and targets for this proof. Let's imagine that the function we want to prove is called `XXX` (without the `$(KYBER_NAMESPACE)` prefix).
+
+Edit the Makefile and update the definition of the following variables:
+
+* HARNESS_FILE - should be `XXX_harness`
+* PROOF_UID - should be `XXX`
+* PROJECT_SOURCES - should the files containing the source code of XXX
+* CHECK_FUNCTION_CONTRACTS - set to the `XXX`, but _with_ the `$(KYBER_NAMESPACE)` prefix if required
```

Comment:
Namespace changed

Reply:
And in multiple places below as well, no longer commenting on each of them.

Reply:
Done